radixdlt/src/main/java/com/radixdlt/consensus/VertexStore.java

```diff
@@ -43,12 +63,48 @@ public void syncToQC(QuorumCertificate qc) {
 		}
 	}
 
-	public void insertVertex(Vertex vertex) {
+	public void insertVertex(Vertex vertex) throws VertexInsertionException {
+		final Vertex parent = vertices.get(vertex.getParentId());
+		if (parent == null) {
+			throw new MissingParentException(vertex.getParentId());
+		}
+
 		this.syncToQC(vertex.getQC());
-		vertices.add(vertex);
+
+		if (vertex.getAtom() != null) {
+			try {
+				this.engine.store(vertex.getAtom());
+			} catch (RadixEngineException e) {
+				throw new VertexInsertionException("Failed to execute", e);
+			}
+		}
+
+		vertices.put(vertex.getId(), vertex);
+	}
+
+	public Vertex commitVertex(Hash vertexId) {
+		Vertex vertex = vertices.get(vertexId);
+		if (vertex == null) {
+			throw new IllegalStateException("Committing a vertex which was never inserted: " + vertexId);
+		}
+
+		committedVertices.put(vertexId, vertex);
+		return vertex;
+	}
+
+	public List<Vertex> getPathFromRoot(Hash vertexId) {
+		List<Vertex> path = new ArrayList<>();
+
+		Vertex vertex = vertices.get(vertexId);
+		while (!committedVertices.containsKey(vertex.getId())) {
```

Comment:
Should probably be `while (vertex != null && !committedVertices.containsKey(vertex.getId())) {` to avoid NPE on arbitrary hash being passed in as argument.